### PR TITLE
Fix uninstall not working when version changes.

### PR DESCRIPTION
**Pull request by reitermarkus**
### Changes to the core
- [x] Followed [hacking.md](https://github.com/caskroom/homebrew-cask/blob/master/doc/development/hacking.md).

---
### Todo
- [x] Add test for message when other versions remain installed.
- [x] Add test for uninstalling multiple versions consecutively.

---

https://github.com/caskroom/homebrew-cask/issues/2988
https://github.com/caskroom/homebrew-cask/issues/14058

### CODE REVIEW

File: lib/hbc/cask.rb [hunk 1]
```diff
@@ -72,11 +72,21 @@ def timestamped_versions
   end
 
   def versions
-    timestamped_versions.map(&:first).uniq
+    timestamped_versions.map(&:first)
+                        .reverse
+                        .uniq
+                        .reverse
```
Comment (jawshooah) [inline lib/hbc/cask.rb]: Why reverse twice? Isn't `.uniq.reverse` sufficient?

Comment (reitermarkus) [inline lib/hbc/cask.rb]: Consider this:

```ruby
timestamped_versions = [['a', '1'], ['b', '2'], ['a', '3']]

timestamped_versions.map(&:first) --> ['a', 'b', 'a']
                    .uniq         --> ['a', 'b']
```

Now, `.last` doesn't actually return the version which was installed last.

Comment (jawshooah) [inline lib/hbc/cask.rb]: Ah, gotcha.

**Comment by jawshooah** [comment]
I think a change of this magnitude warrants some new tests. Once those are added and the currently failing tests are fixed, I'm all for merging this 👍

**Comment by reitermarkus** [comment]
@jawshooah, what exactly is `PERSISTENT_METADATA_SUBDIRS`? Is the `gpg` sub-directory used for anything else than `signature.asc`?

According to `fetch_sig` in `gpg.rb`, this is only used for caching. Wouldn't  `Hbc.cache` be a better place for `signature.asc` files?

**Comment by jawshooah** [comment]
@reitermarkus I'm not sure. Looks like it was added in https://github.com/caskroom/homebrew-cask/pull/6117 "to provide persistent per-distribution locations to for GPG data". I'd be fine with moving the signature file to `Hbc.cache`

**Comment by reitermarkus** [comment]
Ah, I had an error in thinking:

I thought the `gpg` directory would be in`caskroom/cask/.metadata/gpg`, which would result in a conflict when globbing the `.metadata` directory to retreive timestamped versions.

But actually it's stored in `caskroom/cask/.metadata/version/timestamp/gpg`, so there's no conflict.

**Comment by reitermarkus** [comment]
@jawshooah, could you take a look at the test I have added (and the one I have altered)? 

Also, not sure what would be the best way to reuse the code from the `uninstalls one version at a time`. `MiniTest` doesn't have a `before(:each)`, right? 

Or should I simply add the test for the `cask 1.2.3 is still installed` message into the existing one?

**Comment by reitermarkus** [comment]
Ok, apparently [`after` and `before`](http://ruby-doc.org/stdlib-2.0.0/libdoc/minitest/rdoc/MiniTest/Spec.html#method-c-after) are equivalent to `teardown` and `setup`. Strange this is, `after` and `before` are working, `teardown` and `setup` are not.

**Comment by jawshooah** [comment]
@reitermarkus LGTM, merge when ready 👍

**Comment by reitermarkus** [comment]
This pull request was merged in 36a98c5a4639f62cad6ecd82f896cbc080c0969e...1835d6d5bd415da0d32924416e2e6edb8fa585b7.

**Pull request closed without merging**